varname/ignore.py

```diff
@@ -0,0 +1,360 @@
+        if module and module != self.module:
+            return False
+
+        if (
+                not module and
+                not frame_matches_module_by_ignore_id(frame, self.module)
+        ):
+            return False
+
+        source = Source.for_frame(frame)
+        check_qualname_by_source(source, self.module.__name__, self.qualname)
+
+        return fnmatch(
+            source.code_qualname(frame.f_code),
+            self.qualname
+        )
+
+    def __repr__(self) -> str:
+        return ('<IgnoreModuleQualname'
+                f'({self.module.__name__!r}, {self.qualname})>')
+
+class IgnoreFilenameQualname(IgnoreElem, attrs=['filename', 'qualname']):
+    """Ignore calls with given qualname in the module with the filename"""
+
+    # check qualname uniqueness in this as well?
+    # But where to find module.__dict__ for Source.for_filename?
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+
+        return (
+            path.realpath(frame.f_code.co_filename) ==
+            path.realpath(self.filename) and
+            fnmatch(
+                Source.for_frame(frame).code_qualname(frame.f_code),
+                self.qualname
+            )
+        )
+
+    def __repr__(self) -> str:
+        return f'<IgnoreFilenameQualname({self.filename!r}, {self.qualname})>'
+
+class IgnoreOnlyQualname(IgnoreElem, attrs=['_none', 'qualname']):
+    """Ignore calls that match the given qualname, across all frames."""
+
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        frame = frameinfos[frame_no].frame
+
+        # module is None, check qualname only
+        return fnmatch(
+            Source.for_frame(frame).code_qualname(frame.f_code),
+            self.qualname
+        )
+
+    def __repr__(self) -> str:
+        return f'<IgnoreOnlyQualname({self.qualname!r})>'
+
+def create_ignore_elem(ignore_elem: IgnoreElemType) -> IgnoreElem:
+    """Create an ignore element according to the type"""
+    if isinstance(ignore_elem, ModuleType):
+        return IgnoreModule(ignore_elem)
+    if isinstance(ignore_elem, str):
+        return (IgnoreDirname(ignore_elem)
+                if path.isdir(ignore_elem)
+                else IgnoreFilename(ignore_elem))
+    if hasattr(ignore_elem, '__code__'):
+        return IgnoreFunction(ignore_elem)
+    if not isinstance(ignore_elem, tuple) or len(ignore_elem) != 2:
+        raise ValueError(f'Unexpected ignore item: {ignore_elem!r}')
+    # is tuple and len == 2
+    if hasattr(ignore_elem[0], '__code__') and isinstance(ignore_elem[1], int):
+        return IgnoreDecorated(ignore_elem)
+    if isinstance(ignore_elem[0], ModuleType):
+        return IgnoreModuleQualname(ignore_elem)
+    if isinstance(ignore_elem[0], str):
+        return IgnoreFilenameQualname(ignore_elem)
+    if ignore_elem[0] is None:
+        return IgnoreOnlyQualname(ignore_elem)
```

Comment:
check that ignore_elem[1] is a string.

Also I'd suggest either changing all these cases where ignore_elem is a tuple to `Ignore<subclass name>(*ignore_elem)` and then you should be able to simplify `subclass_init`, or you should probably update the `__repr__` of these classes to reflect that they take a tuple, not two arguments. You generally want `eval(repr(x))` to be the same as `x` where possible. On that note, why are the reprs surrounded by `<>`? You could also probably just have one `__repr__` in the base class, something like `f"{self.__class__.name}({', '.join(repr(getattr(attr_value, '__name__', attr_value)) for attr_value in attr_values)})"`

Reply:
These `__repr__`s are mostly used in debug. `<>` is used to separate it from other messages. But I am not sure if that has a special meaning in `__repr__`. Your other suggestions in this one look good to me.